File: ai-crm/backend/setu/sovereign_routing_adapter.py
REQUIRED_FIELDS = [
    "execution_id",
    "trace_id",
    "source_system",
    "actor",
    "intent_type",
    "target_system",
    "parameters",
    "priority",
    "timestamp",
    "schema_version",
    "tenant_id"
]

REQUIRED_GATED_FIELDS = [
    "status",
    "attestation_id",
    "policy_id",
    "policy_version",
    "checked_at"
]


def _assert_required(execution):
    missing = [field for field in REQUIRED_FIELDS if execution.get(field) is None]
    if missing:
        raise ValueError("Missing execution fields: " + ", ".join(missing))


def _default_gated_bridge_validator(execution):
    gated = (execution.get("governance") or {}).get("gated_bridge")
    if not gated:
        return {"ok": False, "reason": "gated_bridge_missing"}

    missing = [field for field in REQUIRED_GATED_FIELDS if gated.get(field) is None]
    if missing:
        return {"ok": False, "reason": "gated_bridge_incomplete", "missing_fields": missing}

    if gated.get("status") != "approved":
        return {"ok": False, "reason": "gated_bridge_not_approved", "status": gated.get("status")}

    return {"ok": True, "governance": {"gated_bridge": gated}}


def _build_sarathi_payload(execution):
    return {
        "sarathi_version": "1.0",
        "execution_id": execution.get("execution_id"),
        "trace_id": execution.get("trace_id"),
        "tenant_id": execution.get("tenant_id"),
        "intent_type": execution.get("intent_type"),
        "source_system": execution.get("source_system"),
        "target_system": execution.get("target_system"),
        "parameters": execution.get("parameters"),
        "priority": execution.get("priority"),
        "timestamp": execution.get("timestamp"),
        "schema_version": execution.get("schema_version"),
        "actor": execution.get("actor")
    }


def _build_bhiv_envelope(execution, sarathi_payload):
    return {
        "envelope_version": "1.0",
        "execution": {
            "execution_id": execution.get("execution_id"),
            "trace_id": execution.get("trace_id"),
            "tenant_id": execution.get("tenant_id"),
            "intent_type": execution.get("intent_type"),
            "source_system": execution.get("source_system"),
            "target_system": execution.get("target_system"),
            "parameters": execution.get("parameters"),
            "priority": execution.get("priority"),
            "timestamp": execution.get("timestamp"),
            "schema_version": execution.get("schema_version"),
            "actor": execution.get("actor")
        },
        "routing": sarathi_payload,
        "governance": execution.get("governance"),
        "provenance": execution.get("provenance"),
        "replay": execution.get("replay")
    }


class SovereignRoutingAdapter:
    def __init__(self, gated_bridge_validator=None):
        self.gated_bridge_validator = gated_bridge_validator or _default_gated_bridge_validator

    def to_sarathi_payload(self, execution):
        _assert_required(execution)
        return _build_sarathi_payload(execution)

    def to_bhiv_envelope(self, execution):
        _assert_required(execution)
        sarathi_payload = _build_sarathi_payload(execution)
        return _build_bhiv_envelope(execution, sarathi_payload)

    def build_routing_packet(self, execution):
        try:
            _assert_required(execution)
        except ValueError as error:
            return {
                "ok": False,
                "reason": "execution_contract_invalid",
                "details": str(error)
            }

        gated = self.gated_bridge_validator(execution)
        if not gated.get("ok"):
            return {
                "ok": False,
                "reason": gated.get("reason"),
                "details": gated.get("missing_fields") or gated.get("status")
            }

        sarathi_payload = _build_sarathi_payload(execution)
        bhiv_envelope = _build_bhiv_envelope(execution, sarathi_payload)

        return {
            "ok": True,
            "sarathi_payload": sarathi_payload,
            "bhiv_envelope": bhiv_envelope,
            "governance": gated.get("governance")
        }

File: ai-crm/integration/sovereign_routing_adapter.js
const REQUIRED_FIELDS = [
  'execution_id',
  'trace_id',
  'source_system',
  'actor',
  'intent_type',
  'target_system',
  'parameters',
  'priority',
  'timestamp',
  'schema_version',
  'tenant_id'
];

const REQUIRED_GATED_FIELDS = [
  'status',
  'attestation_id',
  'policy_id',
  'policy_version',
  'checked_at'
];

const assertRequired = (execution) => {
  const missing = REQUIRED_FIELDS.filter((field) => execution?.[field] === undefined || execution?.[field] === null);
  if (missing.length > 0) {
    throw new Error(`Missing execution fields: ${missing.join(', ')}`);
  }
};

const defaultGatedBridgeValidator = (execution) => {
  const gated = execution?.governance?.gated_bridge;
  if (!gated) {
    return { ok: false, reason: 'gated_bridge_missing' };
  }

  const missing = REQUIRED_GATED_FIELDS.filter((field) => gated[field] === undefined || gated[field] === null);
  if (missing.length > 0) {
    return { ok: false, reason: 'gated_bridge_incomplete', missing_fields: missing };
  }

  if (gated.status !== 'approved') {
    return { ok: false, reason: 'gated_bridge_not_approved', status: gated.status };
  }

  return { ok: true, governance: { gated_bridge: gated } };
};

const buildSarathiPayload = (execution) => ({
  sarathi_version: '1.0',
  execution_id: execution.execution_id,
  trace_id: execution.trace_id,
  tenant_id: execution.tenant_id,
  intent_type: execution.intent_type,
  source_system: execution.source_system,
  target_system: execution.target_system,
  parameters: execution.parameters,
  priority: execution.priority,
  timestamp: execution.timestamp,
  schema_version: execution.schema_version,
  actor: execution.actor
});

const buildBhivEnvelope = (execution, sarathiPayload) => ({
  envelope_version: '1.0',
  execution: {
    execution_id: execution.execution_id,
    trace_id: execution.trace_id,
    tenant_id: execution.tenant_id,
    intent_type: execution.intent_type,
    source_system: execution.source_system,
    target_system: execution.target_system,
    parameters: execution.parameters,
    priority: execution.priority,
    timestamp: execution.timestamp,
    schema_version: execution.schema_version,
    actor: execution.actor
  },
  routing: sarathiPayload,
  governance: execution.governance || null,
  provenance: execution.provenance || null,
  replay: execution.replay || null
});

export const createSovereignRoutingAdapter = (config = {}) => {
  const { validateGatedBridge = defaultGatedBridgeValidator } = config;

  return {
    validateGatedBridge,
    toSarathiPayload: (execution) => {
      assertRequired(execution);
      return buildSarathiPayload(execution);
    },
    toBhivEnvelope: (execution) => {
      assertRequired(execution);
      const sarathiPayload = buildSarathiPayload(execution);
      return buildBhivEnvelope(execution, sarathiPayload);
    },
    buildRoutingPacket: (execution) => {
      try {
        assertRequired(execution);
      } catch (error) {
        return {
          ok: false,
          reason: 'execution_contract_invalid',
          details: error?.message || 'missing required fields'
        };
      }

      const gated = validateGatedBridge(execution);
      if (!gated.ok) {
        return {
          ok: false,
          reason: gated.reason,
          details: gated.missing_fields || gated.status || null
        };
      }

      const sarathiPayload = buildSarathiPayload(execution);
      const bhivEnvelope = buildBhivEnvelope(execution, sarathiPayload);

      return {
        ok: true,
        sarathi_payload: sarathiPayload,
        bhiv_envelope: bhivEnvelope,
        governance: gated.governance
      };
    }
  };
};
